docs(cn): translate Reference/react/separating events from effects   into Chinese (#1167)
reactjs/zh-hans.react.dev: src/content/learn/separating-events-from-effects.md

---
title: 将事件从 Effect 中分开
---

<Intro>

> 事件处理函数只有在你再次执行同样的交互时才会重新运行。Effect 和事件处理函数不一样，它只有在读取的 prop 或 state 值和上一次渲染不一样时才会重新同步。有时你需要这两种行为的混合体：即一个 Effect 只在响应某些值时重新运行，但是在其他值变化时不重新运行。本章将会教你怎么实现这一点。

Reviewer: ```suggestion
事件处理函数只有在你再次执行同样的交互时才会重新运行。Effect 和事件处理函数不一样，它只有在读取的 props 或 state 值和上一次渲染不一样时才会重新同步。有时你需要这两种行为的混合体：即一个 Effect 只在响应某些值时重新运行，但是在其他值变化时不重新运行。本章将会教你怎么实现这一点。
```

</Intro>

<YouWillLearn>

- 怎么在事件处理函数和 Effect 之间做选择 

> - 为什么 Effects 是响应式的，而事件处理函数不是

Reviewer: Effect 应该考虑统一用单数

- 当你想要 Effect 的部分代码变成非响应式时要做些什么
- Effect Event 是什么，以及怎么从 Effect 中提取
- 怎么使用 Effect Event 读取最新的 props 和 state

</YouWillLearn>

## 在事件处理函数和 Effect 中做选择 {/*choosing-between-event-handlers-and-effects*/}

首先让我们回顾一下事件处理函数和 Effect 的区别。

假设你正在实现一个聊天室组件，需求如下：

1. 组件应该自动连接选中的聊天室。

> 1. 每当你点击 “Send” 按钮，组件应该在当前聊天界面发送一条消息。

Reviewer: 此处以及后面的 “‘Send 按钮’” ，烦请考虑（1）能否改为“‘发送’按钮”（2）如果确实使用英文更何时，看看是否应该统一大小写

Reviewer: （3）如果使用英文，是否需要使用引号？
> 后文似乎有地方没有用引号，但是这几处使用了

Author: 大小写已统一，是否使用引号是跟随原文的，不确定是否应该修改

Reviewer: 无论原文是否使用引号，翻译为中文时都建议统一格式，统一为“‘Send’ 按钮”，或“Send 按钮”。

> 假设你已经实现了这部分代码，但是还没有确定应该放在哪里。你是应该用事件处理函数还是 Effect 呢？每当你需要回答这个问题时，请考虑一下 [**为什么** 代码需要运行](/learn/synchronizing-with-effects#what-are-effects-and-how-are-they-different-from-events)。

Reviewer: ```suggestion
假设你已经实现了这部分代码，但是还没有确定应该放在哪里。你是应该用事件处理函数还是 Effect 呢？每当你需要回答这个问题时，请考虑一下 [为什么代码需要运行](/learn/synchronizing-with-effects#what-are-effects-and-how-are-they-different-from-events)。
```

    });
    connection.connect();
    return () => connection.disconnect();
  }, [roomId]); // ✅ 声明所有依赖项
  // ...
```

> 这个方法解决了问题。注意你必须从 Effect 依赖项中 **移除** `onConnected`。***Effect Event 是非响应式的并且必须从依赖项中删除**。

Reviewer 2: 这里多了个星号

}
```

> 稍后向你的站点添加多个路由。现在 `Page` 组件接收一个包含当前路径的 `url` prop。你想把 `url` 作为 `logVisit` 调用的一部分进行传递，但是依赖项检查工具控诉：

Reviewer: ```suggestion
稍后向你的站点添加多个路由。现在 `Page` 组件接收包含当前路径的 `url` props。你想把 `url` 作为 `logVisit` 调用的一部分进行传递，但是依赖项检查工具会提示：
```

</Sandpack>

> 这不意味着 `useEffectEvent` **总是** 正确的解决方案。你只能把它用在你不需要变成响应式的代码上。上面的 sandbox 中，你不需要 Effect 的代码响应 `canMove`。这就是提取 Effect Event很有意义的原因。

Reviewer 2: “Event很有意义的原因” 少了个空格

  useEffect(() => {
    const id = setInterval(() => {
      onTick(); // ✅ Good: 只在 Effect 内部局部调用
    }, delay);
    return () => {
      clearInterval(id);
    };
  }, [delay]); // 不需要指定 “onTick” (Effect Event) 作为依赖项
}
```

Effect Event 是 Effect 代码的非响应式“片段”。他们应该在使用他们的 Effect 的旁边。

<Recap>

- 事件处理函数在响应特定交互时运行。

> - Effect在需要同步的时候运行。

Reviewer: ```suggestion
- Effect 在需要同步的时候运行。
```

- 事件处理函数内部的逻辑是非响应式的。
- Effect 内部的逻辑是响应式的。
- 你可以将非响应式逻辑从 Effect 移到 Effect Event 中。
- 只在 Effect 内部调用 Effect Event。
- 不要将 Effect Event 传给其他组件或者 Hook。

</Recap>

<Challenges>

#### 修复一个不更新的变量 {/*fix-a-variable-that-doesnt-update*/}

`Timer` 组件保存了一个 `count` 的 state 变量，这个变量每秒增加一次。每次增加的值存储在 `increment` state 变量中。你可以使用加减按钮控制 `increment` 变量。

> 但是无论你点击加号按钮多少次，计数器每秒都只增加 １。这段代码存在什么问题呢？为什么 Effect 内部的 `increment` 总是等于 `1` 呢？找出错误并修复它。

Reviewer: ```suggestion
但是无论你点击加号按钮多少次，计数器每秒都只增加 。这段代码存在什么问题呢？为什么 Effect 内部的 `increment` 总是等于 `1` 呢？找出错误并修复它。
```

Author: 这里去掉1好像不是很通顺》

Reviewer: 有点尴尬，时间太久远，不太记得当时什么想的。现在看起来可能是我评论错了。

</Sandpack>

现在当 `increment` 变化时，React 会重新同步你的 Effect，这会重启 interval。

</Solution>

#### 修复一个冻结的计数器 {/*fix-a-freezing-counter*/}

`Timer` 组件保存了一个 `count` 的 state 变量，这个变量每秒增加一次。每次增加的值存储在 `increment` state 变量中，你可以使用加减按钮控制它。例如，尝试点击加号按钮九次，注意现在 `count` 每次都增加 10 而不是 1。

这个用户接口有一个小问题。你可能注意到如果你每秒内按压加减按钮不止一次， 那计时器本身似乎就会暂停。它只在你最后一次按压按钮的一秒后恢复。找出为什么会发生这种现象，并修复它以便计时器能 **每** 秒滴答作响而不中断。

<Hint>

> 似乎设置计时器的 Effect 对 `increment` 值的变化做出了 “响应”。为了调用 `setCount` 而使用当前 `increment` 值的代码行真的需要是响应式吗？

Reviewer: ```suggestion
似乎设置计时器的 Effect 对 `increment` 值的变化做出了“响应”。为了调用 `setCount` 而使用当前 `increment` 值的代码行真的需要是响应式吗？
```

<Solution>

> 在 Effect Event 内部，`roomId` 是　**Effect Event 被调用时刻** 的值。

Reviewer 2: 这里“是”的后面多了个空格还是空格变成全角了？
在预览时看到间隔很长！
![](https://github.com/reactjs/zh-hans.react.dev/assets/30925732/9cb78d02-2825-4186-9da5-6b223363a211)
